src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -85,35 +91,45 @@ private static string GetSchema(ResourceScope targetScope)
 
             return "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#";
         }
+
         private readonly EmitterContext context;
         private readonly EmitterSettings settings;
+        private readonly IDictionary<string, IDictionary<int, IList<(int start, int end)>>> rawSourceMap;
+
+        public Dictionary<int, (string, int)>? SourceMap; // ARM JSON line => (Bicep File, Bicep Line)
 
         public TemplateWriter(SemanticModel semanticModel, EmitterSettings settings)
         {
             this.context = new EmitterContext(semanticModel, settings);
             this.settings = settings;
+            this.rawSourceMap = new Dictionary<string, IDictionary<int, IList<(int, int)>>>();
         }
 
         public void Write(JsonTextWriter writer)
         {
-            // Template is used for calcualting template hash, template jtoken is used for writing to file.
+            // Template is used for calculating template hash, template JToken is used for writing to file.
             var (template, templateJToken) = GenerateTemplateWithoutHash();
+
             var templateHash = TemplateHelpers.ComputeTemplateHash(template.ToJToken());
             if (templateJToken.SelectToken(GeneratorMetadataPath) is not JObject generatorObject)
             {
                 throw new InvalidOperationException($"generated template doesn't contain a generator object at the path {GeneratorMetadataPath}");
             }
-            generatorObject.Add("templateHash", templateHash);
+            generatorObject.Add(TemplateHashPropertyName, templateHash);
+
+            if (this.context.Settings.EnableSourceMapping)
```

Comment:
Rather than storing state in the `TemplateWriter` class, could we return this as an output for the `Write` function?

Reply:
In order to do this, the ITemplateWriter interface would need to be updated--which results in uglier/throwaway code for the other implementing classes. Any thoughts around this?